diff -urN aa/allegro-4.1.14/configure.in allegro-4.1.14/configure.in
--- aa/allegro-4.1.14/configure.in	2004-05-23 21:23:07.930046696 +0200
+++ allegro-4.1.14/configure.in	2004-05-23 21:21:51.225707528 +0200
@@ -245,7 +245,7 @@
   COMPILE_PROGRAM=COMPILE_PROFILE
   COMPILE_CXX_PROGRAM=COMPILE_CXX_PROFILE
   COMPILE_S_PROGRAM=COMPILE_S_PROFILE
-  PROG_LDFLAGS="-pg \$(LDFLAGS)"
+  PROG_LDFLAGS="-fno-omit-frame-pointer -pg \$(LDFLAGS)"
   LIB_TO_LINK=allp
   PLUGIN_LIB=lib/unix/libapdat.a
   allegro_build_profiling_library=yes
@@ -671,7 +671,7 @@
      ALLEGRO_DEBUG_CFLAGS="$ALLEGRO_DEBUG_CFLAGS -DDMALLOC"
   fi
   CFLAGS="$CFLAGS $TARGET_ARCH"
-	ALLEGRO_PROFILE_CFLAGS="$CFLAGS -pg $WFLAGS"
+	ALLEGRO_PROFILE_CFLAGS="$CFLAGS -fno-omit-frame-pointer -pg $WFLAGS"
   ALLEGRO_SFLAGS="-x assembler-with-cpp"
   ALLEGRO_SHAREDLIB_CFLAGS="-fPIC -DALLEGRO_SHARED"
 else
