diff -ruN allegro-4.1.11./misc/allegro-config.in allegro-4.1.11/misc/allegro-config.in
--- allegro-4.1.11./misc/allegro-config.in	2003-06-30 01:27:31.000000000 +0200
+++ allegro-4.1.11/misc/allegro-config.in	2003-08-25 00:50:30.000000000 +0200
@@ -21,7 +21,7 @@
 
 accepts_frameworks=no
 
-allegro_ldflags="@LDFLAGS@"
+allegro_ldflags=""
 allegro_libs="@LIBS@"
 allegro_frameworks="@FRAMEWORKS@"
 allegro_cflags=""
