--- allegro-4.4.1.1/misc/allegro-config.in~	2010-03-03 11:07:40.000000000 +0100
+++ allegro-4.4.1.1/misc/allegro-config.in	2010-03-15 16:01:04.957931016 +0100
@@ -190,9 +190,9 @@
 if test "$echo_libs" = "yes"; then
    test -z "$addon_form" && libdirs=-L${lib_path}
    if test "$static_libs" = "yes"; then
-      echo $libdirs $allegro_ldflags -l${lib_type} $allegro_libs
+      echo $libdirs$allegro_ldflags -l${lib_type} $allegro_libs
    else
-      echo $libdirs $allegro_ldflags -l${lib_type}
+      echo $libdirs$allegro_ldflags -l${lib_type}
    fi
 fi
 
