--- allegro-4.9.3/misc/allegro5-config.in~	2008-05-12 13:17:40.502753489 +0200
+++ allegro-4.9.3/misc/allegro5-config.in	2008-05-12 13:18:32.301896453 +0200
@@ -185,9 +185,9 @@
 if test "$echo_libs" = "yes"; then
    libdirs=-L$lib_path
    if test "$static_libs" = "yes"; then
-      echo $libdirs $allegro_ldflags -l$lib_type $allegro_libs
+      echo $libdirs$allegro_ldflags -l${lib_type} $allegro_libs
    else
-      echo $libdirs $allegro_ldflags -l$lib_type-$version
+      echo $libdirs$allegro_ldflags -l${lib_type}-${version} -l${lib_type}_unsharable
    fi
 fi
 
