--- allegro-4.0.2/makefile.in.orig	Fri Jul  5 17:48:43 2002
+++ allegro-4.0.2/makefile.in	Fri Jul  5 17:52:29 2002
@@ -392,8 +392,9 @@
 
 install-programs: programs
 	$(mkinstalldirs) $(bindir)
-	@for prog in $(ALLEGRO_LIB_EXES) $(ALLEGRO_DATAFILE_EXES); do \
-	  if test "X`echo $$prog | sed -e s,tools/.\*,,`" = X; then \
+	@for prog in $(ALLEGRO_LIB_EXES) $(ALLEGRO_DATAFILE_EXES) \
+	  $(ALLEGRO_EXAMPLE_EXES) ; do \
+	  if test "X`echo $$prog | sed -e s,.*/.\*,,`" = X; then \
 	    echo Installing $$prog to $(bindir); \
 	    $(INSTALL_PROGRAM) $$prog $(bindir); \
 	  fi; \
@@ -465,9 +466,10 @@
 	rm -rf $(includedir)/allegro
 
 uninstall-programs:
-	@for prog in $(ALLEGRO_LIB_EXES) $(ALLEGRO_DATAFILE_EXES); do \
-	  if test "X`echo $$prog | sed -e s,tools/.\*,,`" = X; then \
-	    p=`echo $$prog | sed -e s,tools/,,`; \
+	@for prog in $(ALLEGRO_LIB_EXES) $(ALLEGRO_DATAFILE_EXES) \
+	  $(ALLEGRO_EXAMPLE_EXES) ; do \
+	  if test "X`echo $$prog | sed -e s,.*/.\*,,`" = X; then \
+	    p=`echo $$prog | sed -e s,,,`; \
 	    echo rm -f $(bindir)/$$p; \
 	    rm -f $(bindir)/$$p; \
 	  fi; \
