--- allegro-4.9.2/configure.in~	2008-03-25 13:48:59.559395652 +0100
+++ allegro-4.9.2/configure.in	2008-03-25 13:58:36.726987554 +0100
@@ -275,7 +275,7 @@
   COMPILE_PROGRAM=COMPILE_DEBUG
   COMPILE_CXX_PROGRAM=COMPILE_CXX_DEBUG
   COMPILE_S_PROGRAM=COMPILE_S_DEBUG
-  PROG_LDFLAGS="-g \$(LDFLAGS)"
+  PROG_LDFLAGS="-fno-omit-frame-pointer -pg \$(LDFLAGS)"
   LIB_TO_LINK=alld
   PLUGIN_LIB=lib/unix/libaddat.a
   allegro_build_debugging_library=yes
@@ -824,7 +824,7 @@
      ALLEGRO_DEBUG_CFLAGS="$ALLEGRO_DEBUG_CFLAGS -DDMALLOC"
   fi
   CFLAGS="$CFLAGS $TARGET_ARCH"
-  ALLEGRO_PROFILE_CFLAGS="$CFLAGS -pg $WFLAGS"
+  ALLEGRO_PROFILE_CFLAGS="$CFLAGS -fno-omit-frame-pointer -pg $WFLAGS"
   ALLEGRO_SFLAGS="-x assembler-with-cpp"
   ALLEGRO_SHAREDLIB_CFLAGS="-fPIC -DALLEGRO_SHARED"
 else
