--- allegro-3.9.38/makefile.in	Thu Sep 13 21:35:37 2001
+++ allegro-3.9.38-p/makefile.in	Tue Nov 13 13:39:34 2001
@@ -267,90 +267,73 @@
 	@echo "Compressed postscript documentation generated: allegro.ps.gz"
 
 install-lib: lib
-	$(mkinstalldirs) $(libdir)
+	$(mkinstalldirs) $(DESTDIR)/$(libdir)
 	@for l in alleg alld allp; do \
 	  if test -f $(LIBDIR)/lib$${l}.a; then \
-	    echo Installing $(LIBDIR)/lib$${l}.a to $(libdir); \
-	    $(INSTALL_DATA) $(LIBDIR)/lib$${l}.a $(libdir)/; \
+	    echo Installing $(LIBDIR)/lib$${l}.a to $(DESTDIR)/$(libdir); \
+	    $(INSTALL_DATA) $(LIBDIR)/lib$${l}.a $(DESTDIR)/$(libdir)/; \
 	  fi; \
 	done
 	@for l in alleg alld allp; do \
 	  if test -f $(LIBDIR)/lib$${l}-$(shared_version).so; then \
-	    echo Installing $(LIBDIR)/lib$${l}-$(shared_version).so to $(libdir); \
-	    $(INSTALL_DATA) $(LIBDIR)/lib$${l}-$(shared_version).so $(libdir)/; \
-	    $(INSTALL_DATA) $(LIBDIR)/lib$${l}_unsharable.a $(libdir)/; \
+	    echo Installing $(LIBDIR)/lib$${l}-$(shared_version).so to $(DESTDIR)/$(libdir); \
+	    $(INSTALL_DATA) $(LIBDIR)/lib$${l}-$(shared_version).so $(DESTDIR)/$(libdir)/; \
+	    $(INSTALL_DATA) $(LIBDIR)/lib$${l}_unsharable.a $(DESTDIR)/$(libdir)/; \
 	  fi; \
 	done
 	@if test -n "$(ALLEGRO_MODULE_TARGETS)"; then \
-	  $(mkinstalldirs) $(moduledir); \
+	  $(mkinstalldirs) $(DESTDIR)/$(moduledir); \
 	  for m in $(ALLEGRO_MODULE_TARGETS) modules.lst; do \
-	   echo Installing $$m to $(moduledir); \
-	   $(INSTALL_DATA) $$m $(moduledir)/; \
+	   echo Installing $$m to $(DESTDIR)/$(moduledir); \
+	   $(INSTALL_DATA) $$m $(DESTDIR)/$(moduledir)/; \
 	  done; \
 	fi
-	$(mkinstalldirs) $(bindir)
-	@echo Installing allegro-config to $(bindir)
-	@$(INSTALL_PROGRAM) allegro-config $(bindir)
-	@if test -n "$(LIBDIR)/liball*-$(shared_version).so"; then \
-		grep -q $(libdir) /etc/ld.so.conf && ($(LDCONFIG) $(libdir) || true) || (\
-			echo "" ; \
-			echo "Warning: shared library destination $(libdir) is not in ldd search path." ; \
-			echo "Unless you add it to /etc/ld.so.conf, you must set LD_LIBRARY_PATH to include"; \
-			echo "$(libdir) each time you want to run an Allegro program." ; \
-		) ; \
-	fi
+	$(mkinstalldirs) $(DESTDIR)/$(bindir)
+	@echo Installing allegro-config to $(DESTDIR)/$(bindir)
+	@$(INSTALL_PROGRAM) allegro-config $(DESTDIR)/$(bindir)
 
 install-headers:
-	$(mkinstalldirs) $(includedir)/allegro
-	$(INSTALL_DATA) $(srcdir)/include/allegro.h $(includedir)/
-	$(INSTALL_DATA) $(srcdir)/include/xalleg.h $(includedir)/
-	$(INSTALL_DATA) $(srcdir)/include/linalleg.h $(includedir)/
-	$(INSTALL_DATA) $(CONFIG_H) $(includedir)/allegro/
+	$(mkinstalldirs) $(DESTDIR)/$(includedir)/allegro
+	$(INSTALL_DATA) $(srcdir)/include/allegro.h $(DESTDIR)/$(includedir)/
+	$(INSTALL_DATA) $(srcdir)/include/xalleg.h $(DESTDIR)/$(includedir)/
+	$(INSTALL_DATA) $(srcdir)/include/linalleg.h $(DESTDIR)/$(includedir)/
+	$(INSTALL_DATA) $(CONFIG_H) $(DESTDIR)/$(includedir)/allegro/
 	@for file in $(srcdir)/include/allegro/*.h; do \
-	  echo Installing $$file to $(includedir)/allegro; \
-	  $(INSTALL_DATA) $$file $(includedir)/allegro/; \
+	  echo Installing $$file to $(DESTDIR)/$(includedir)/allegro; \
+	  $(INSTALL_DATA) $$file $(DESTDIR)/$(includedir)/allegro/; \
 	done
 
 install-programs: programs
-	$(mkinstalldirs) $(bindir)
+	$(mkinstalldirs) $(DESTDIR)/$(bindir)
 	@for prog in $(ALLEGRO_LIB_EXES) $(ALLEGRO_DATAFILE_EXES); do \
 	  if test "X`echo $$prog | sed -e s,tools/.\*,,`" = X; then \
-	    echo Installing $$prog to $(bindir); \
-	    $(INSTALL_PROGRAM) $$prog $(bindir); \
+	    echo Installing $$prog to $(DESTDIR)/$(bindir); \
+	    $(INSTALL_PROGRAM) $$prog $(DESTDIR)/$(bindir); \
 	  fi; \
 	done
 
 install-info:
 	@echo Installing Allegro info documentation...
-	$(mkinstalldirs) $(infodir)
-	@$(INSTALL_DATA) docs/allegro.info $(infodir)
-	-@if $(SHELL) -c "$(INSTALL_INFO) --version" >/dev/null 2>&1; then \
-	  if $(INSTALL_INFO) --version 2>&1 | grep -q "Debian"; then \
-	    $(INSTALL_INFO) --section Development -- $(infodir)/allegro.info; \
-	  else \
-	    $(INSTALL_INFO) $(infodir)/allegro.info $(INFO_DIR); \
-	  fi; \
-	else \
-	  true; \
-	fi
+	$(mkinstalldirs) $(DESTDIR)/$(infodir)
+	@$(INSTALL_DATA) docs/allegro.info $(DESTDIR)/$(infodir)
 
 install-gzipped-info: install-info
 	@echo Compressing documentation...
-	@gzip -9 -f $(infodir)/allegro.info
+	@gzip -9 -f $(DESTDIR)/$(infodir)/allegro.info
 
 install-man:
-	$(mkinstalldirs) $(mandir)/man3
+	$(mkinstalldirs) $(DESTDIR)/$(mandir)/man3
 	@for file in $(srcdir)/docs/*.3; do \
-	  echo Installing $$file to $(mandir)/man3; \
-	  $(INSTALL_DATA) $$file $(mandir)/man3/; \
+	  echo Installing $$file to $(DESTDIR)/$(mandir)/man3; \
+	  $(INSTALL_DATA) $$file $(DESTDIR)/$(mandir)/man3/; \
 	done
 
 install-gzipped-man:
-	$(mkinstalldirs) $(mandir)/man3
+	$(mkinstalldirs) $(DESTDIR)/$(mandir)/man3
 	@for file in $(srcdir)/docs/*.3; do \
-	  echo Installing $$file.gz to $(mandir)/man3; \
-	  $(INSTALL_DATA) $$file $(mandir)/man3/; \
-	  gzip -9 -f $(mandir)/man3/`echo $$file | sed -e "s:.*/::"`; \
+	  echo Installing $$file.gz to $(DESTDIR)/$(mandir)/man3; \
+	  $(INSTALL_DATA) $$file $(DESTDIR)/$(mandir)/man3/; \
+	  gzip -9 -f $(DESTDIR)/$(mandir)/man3/`echo $$file | sed -e "s:.*/::"`; \
 	done
 
 uninstall: uninstall-lib uninstall-modules uninstall-headers uninstall-programs uninstall-info uninstall-man
