--- allegro-4.9.2/docs/src/allegro._tx~	2008-03-24 20:21:49.962436955 +0100
+++ allegro-4.9.2/docs/src/allegro._tx	2008-03-24 20:22:11.686808839 +0100
@@ -31,7 +31,7 @@
 @# This should remain commented at the moment (dvi output) @$@finalout
 @$
 @$@ifinfo
-@$@dircategory Libraries
+@$@dircategory Libraries:
 @$@direntry
 @$* Allegro: (allegro).			The Allegro game programming library.
 @$@end direntry
