--- allegro-4.1.12/configure.in.orig	2003-11-11 21:59:10.385928944 +0100
+++ allegro-4.1.12/configure.in	2003-11-11 22:02:54.074923048 +0100
@@ -645,41 +645,16 @@
 AC_CHECK_FUNCS(memcmp mkstemp stricmp strlwr strupr)
 
 dnl Tweak header files for library build
-CFLAGS="$CFLAGS -DALLEGRO_LIB_BUILD"
+CFLAGS="$TARGET_ARCH -DALLEGRO_LIB_BUILD"
 
 dnl How to compile C and asm files.
 if test -n "GCC"; then
-  if test "X$allegro_strict_warnings" = "Xyes"; then
-    WFLAGS="-Wall -W -Wstrict-prototypes -Wno-unused -Werror"
-  else
-    WFLAGS="-Wall -Wno-unused"
-  fi
-  if test "X$allegro_cv_processor_type" = "Xi386"; then
-    if test "$allegro_exclusive_optimizations" != none; then
-      TARGET_ARCH="-march=$allegro_exclusive_optimizations"
-    else
-      TARGET_ARCH="-mcpu=$allegro_optimizations"
-    fi
-  else
-    TARGET_ARCH=
-  fi
-  if test "X$allegro_cv_can_use_pipe" = "Xyes"; then
-    CFLAGS="$CFLAGS -pipe"
-  fi
-  dnl Allow to specify additional flags.
-  CFLAGS="$CFLAGS $XCFLAGS"
-  WFLAGS="$WFLAGS $WCFLAGS"
-  dnl In this order...
-  ALLEGRO_DEBUG_CFLAGS="$CFLAGS -g $WFLAGS -DDEBUGMODE"
+  WFLAGS="-Wall -Wno-unused"
+  ALLEGRO_DEBUG_CFLAGS="-g $WFLAGS -DDEBUGMODE"
   if test "X$allegro_debug_with_fortify" = "Xyes"; then
      ALLEGRO_DEBUG_CFLAGS="$ALLEGRO_DEBUG_CFLAGS -DFORTIFY"
   fi
-  ALLEGRO_PROFILE_CFLAGS="$CFLAGS -pg $TARGET_ARCH -O2 -funroll-loops -ffast-math $WFLAGS"
-  if test "X$allegro_cv_support_fomit_frame_pointer" = "Xyes"; then
-    CFLAGS="$CFLAGS $TARGET_ARCH -O2 -funroll-loops -ffast-math -fomit-frame-pointer $WFLAGS"
-  else
-    CFLAGS="$CFLAGS $TARGET_ARCH -O2 -funroll-loops -ffast-math $WFLAGS"
-  fi
+  ALLEGRO_PROFILE_CFLAGS="$CFLAGS -pg $WFLAGS"
   ALLEGRO_SFLAGS="-x assembler-with-cpp"
   ALLEGRO_SHAREDLIB_CFLAGS="-fPIC -DALLEGRO_SHARED"
 else
