--- allegro-4.9.2/makefile.in~	2008-03-25 13:37:50.960766155 +0100
+++ allegro-4.9.2/makefile.in	2008-03-25 13:40:51.949418447 +0100
@@ -539,7 +536,7 @@
 	  $(INSTALL_DATA) $$file $(DESTDIR)$(includedir)/allegro5/platform/; \
 	done
 
-INSTALL_EXES = $(filter tools/%, $(ALLEGRO_LIB_EXES) $(ALLEGRO_DATAFILE_EXES))
+INSTALL_EXES = $(filter tools/% demo/% setup/% tests/% examples/%, $(ALLEGRO_EXE_TARGETS))
 
 install-programs: $(INSTALL_EXES)
 	$(mkinstalldirs) $(DESTDIR)$(bindir)
@@ -627,9 +627,9 @@
 	rm -rf $(includedir)/allegro5
 
 uninstall-programs:
-	@for prog in $(ALLEGRO_LIB_EXES) $(ALLEGRO_DATAFILE_EXES); do \
-	  if test "X`echo $$prog | sed -e s,tools/.\*,,`" = X; then \
-	    p=`echo $$prog | sed -e s,tools/,,`; \
+	  @for prog in $(ALLEGRO_EXE_TARGETS); do \
+	    if test "X`echo $$prog | sed -e s,.*/.\*,,`" = X; then \ 
+	      p=`echo $$prog | sed -e s,,,`; \
 	    echo rm -f $(bindir)/$$p; \
 	    rm -f $(bindir)/$$p; \
 	  fi; \
